File: shared/src/commonMain/kotlin/com/moviesshumtimes/tv/sync/RelayUrls.kt
package com.moviesshumtimes.tv.sync

data class RelayHttpUrl(val base: String, val query: String?)

fun relayHttpUrl(relayUrl: String): RelayHttpUrl? {
    val (httpScheme, rest) = when {
        relayUrl.startsWith("wss://") -> "https://" to relayUrl.removePrefix("wss://")
        relayUrl.startsWith("ws://") -> "http://" to relayUrl.removePrefix("ws://")
        else -> return null
    }
    val hostAndPort = rest.substringBefore('/').substringBefore('?')
    if (hostAndPort.isBlank()) return null
    val query = rest.substringAfter('?', missingDelimiterValue = "").takeIf { it.isNotEmpty() }
    return RelayHttpUrl(base = "$httpScheme$hostAndPort", query = query)
}

File: flutter/lib/sync/relay_urls.dart
class RelayHttpUrl {
  final String base;
  final String? query;

  const RelayHttpUrl({required this.base, this.query});
}

RelayHttpUrl? relayHttpUrl(String relayUrl) {
  String httpScheme;
  String rest;
  if (relayUrl.startsWith('wss://')) {
    httpScheme = 'https://';
    rest = relayUrl.substring('wss://'.length);
  } else if (relayUrl.startsWith('ws://')) {
    httpScheme = 'http://';
    rest = relayUrl.substring('ws://'.length);
  } else {
    return null;
  }

  final hostAndPort = rest.split('/').first.split('?').first;
  if (hostAndPort.trim().isEmpty) return null;

  final queryIndex = rest.indexOf('?');
  final query = queryIndex == -1 ? null : (rest.substring(queryIndex + 1).isEmpty ? null : rest.substring(queryIndex + 1));

  return RelayHttpUrl(base: '$httpScheme$hostAndPort', query: query);
}
